### DISCUSSION TITLE
[2023-07-03] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#7187] [Weekly Demo Day]
## The Component Browser Design
This week I was mostly working on [design](https://github.com/enso-org/design/pull/56). Although I presented the improved Component Browser design the last week, I've received multiple important comments regarding design that needs to be changed. The design presented below is already accepted by James, so no further changes are expected. The most important changes in comparison to last week:

- No "search mode" vs "code mode". The design last week assumed these two modes to co-exist, allowing users to either search for components or type the code. This has a significant problem of not allowing users to gradually become power users. This design change influenced both how the components are visible (visibility of `.` and `_`), how they should be matched, and how AI should be integrated.
- Improved handling of local scope mode and shortcut mode.
- Better UX of node icons and AI hints (less UI movement).
- Many smaller UI/UX improvements, including icons and spacing.

<img width="1438" alt="image" src="https://github.com/enso-org/design/assets/1623053/d5d7b4d6-b7df-4bca-a54a-7c919c26be71">

<img width="654" alt="image" src="https://github.com/enso-org/design/assets/1623053/9a399fc3-2db4-4f33-b172-d5278c902fab">

**Comment by MichaelMauderer** [comment]
* Finished up PR for AI Searcher + some planning for next steps. (Awaiting QA)

https://github.com/enso-org/enso/assets/1428930/f8403533-54ba-4ea5-9d3c-6bdf3cf336b5
* Fixed  "Removing a node does not remove the reference to it from connected nodes" via  #7185

https://github.com/enso-org/enso/assets/1428930/85433a69-03dc-4ad9-9563-717a78ff6ff8

**Comment by farmaazon** [comment]
**Node created "by mouse" (tab/enter shortcut or dragging out an edge) will no longer be adjusted).**

**Fix visualization preview flickering**. Now the visualization preview is hidden only when mouse leaves both output port and preview. Also hid the visualization preview from the chooser, as it looked ugly in nodes displayed over the nodes (or wasn't visible at all).

https://github.com/enso-org/enso/assets/3919101/be509314-a5f7-41c8-be0f-a7c46deca477

**Improved handling of mouse hovering events**. Now there is a guarantee that for each "mouse enter" event there will be "mouse leave" event, so tracking hovering state of my component should be now quite easy. We also handle the situations where the shape was moved to another parent (and thus stopped being hovered) - the moved shape will get "leave" event, but the parent's will not if they continue to be hovered.

https://github.com/enso-org/enso/assets/3919101/69b07ede-bb89-4e0b-ad16-977cb08dd213

Other activities include doing a review of new component browser design, discussing it, and making a list of tasks needed to implement a first version of it.

**Comment by vitvakatu** [comment]
Resizing visualizations in QA: https://github.com/enso-org/enso/pull/7164

- Visualizations are aligned to the left border of the node now.
- Default visualization width now equals to the node's width (default height is the same)
- Changing the width of the node also changes visualization width, but only if no manual drag-resizing was applied
- The above condition is reset when reopening the visualization, so that it again starts following the width of the node
- No visual indication that resizing is possible exists
- Two custom shapes in visualization container were replaced with magic Rectangles

https://github.com/enso-org/enso/assets/6566674/2f2525e8-cf10-4c92-953a-b69eb97a954a

**Comment by mwu-tow** [comment]
Weekly update:
* landed https://github.com/enso-org/enso/pull/7121:fixing undo-redo after opening the project;
* landed https://github.com/enso-org/enso/pull/7175: not auto-starting Enso after the installation;
* refactored https://github.com/enso-org/enso/pull/6582 as per CR / discussion over the call, awaiting re-review;
* did some research regarding the EV certificate for code signing, it seems that HW token might be avoided after all;
* observed CI, as always: some strange issues regarding GH-hosted macOS runners.

As for Undo/Redo: I finally figured out what is going on regarding the visualization toggling undo/redo. The issue is non-trivial, I'll need to write this down and get some insight from someone better versed in the visualization's FRPs.

**Comment by hubertp** [comment]
Backend changes relevant for GUI:
- sending function schema for function values #7168 - should help with partially applied functions
- sending `executionComplete` only when execution really succeeds and `executionFailed` when it failed #7143 - should help with managing notifications
- provide method call info for infix operators #7090
- drop-down widgets for extensions functions #7115

**Comment by kazcw** [comment]
New focus APIs; a focused child can handle some events, while allowing others to pass up the focus hierarchy. Used to implement CB focus behavior:

[vokoscreenNG-2023-06-29_10-55-00.webm](https://github.com/enso-org/enso/assets/1047859/f1d9d07c-8c32-4482-ba32-15b6e4e20ae7)

API changes:
- `#[derive(display::Object)]`
- Introduce display object *focus receiver* concept. Many components, when receiving focus, should actually be focused indirectly by focusing a descendant.
    - For example, when the CB Panel receives focus, its descendant at `self.model().grid.model().grid` should be focused, because that's the underlying Grid View, which has its own event handlers. By allowing each level of the hierarchy to define a `focus_receiver`, focus can reach the right object without the CB panel having to know structural details of its descendants.
- New `View::focused_shortcuts()` are delivered to focused children first, and stop bubbling when handled. This replaces `View::default_shortcuts()` with a `focused` condition, when we port to the new focus API everywhere.

This week: I've been working on fuzzy matching.

**Comment by Frizi** [comment]
- Continuing work on #6552 - fixed mouse hover and click handling for node selection and output shape. Fixed layer ordering between node and output shapes. Currently in progress of rewritting node output shape to use only rectangle shapes.
- Improved rectangle shape capabilities to allow more border configurations, as well as optimized its shader output. https://github.com/enso-org/enso/pull/7188

<img width="657" alt="image" src="https://github.com/enso-org/enso/assets/919491/40e5963d-0717-4662-abf4-b9687aa921ed">

**Comment by JanBeelte** [comment]
### Working on our AI Investor Demo.
Main feature that we want to demo is how more complex programs featuring multiple nodes can be generated.
Given the UI and Searcher Designs are still under heavy development we found an easy workaround:
Our AI module will generate multiline Enso-Code and modify the source-file directly on disk.
We will implement hot-reloading in the Language-Server so that the UI picks up these changes.

_Side note:
I am well aware that this is a very hacky solution but I still see value in it even past our investor demo. Using this mechanism we can iterate much faster on the AI code generation and are not relying so much on the IDE Front-End. As soon as the AI results are "stable enough" and the UI changes have been implemented I will work on the proper IDE integration (which should be much simpler by that time)._
